spec/datadog/tracing/client_ip_spec.rb

```diff
@@ -0,0 +1,344 @@
+require 'spec_helper'
+
+require 'datadog/core/header_collection'
+require 'datadog/tracing/client_ip'
+require 'datadog/tracing/metadata/ext'
+
+RSpec::Matchers.define :be_valid_ip do
+  match do |actual|
+    normalised = Datadog::Tracing::ClientIp.strip_decorations(actual)
+    begin
+      Datadog::Tracing::ClientIp.validate_ip(normalised)
+
+      true
+    rescue Datadog::Tracing::ClientIp::InvalidIpError
+      false
+    end
+  end
+end
+
+RSpec.describe Datadog::Tracing::ClientIp do
+  subject(:client_ip) { described_class }
+  let(:tagging_enabled) { true }
+  let(:ip_header_name) { nil }
+
+  before do
+    Datadog.configure do |c|
+      c.tracing.client_ip.disabled = !tagging_enabled
+      c.tracing.client_ip.header_name = ip_header_name
+    end
+  end
+
+  after do
+    without_warnings { Datadog.configuration.reset! }
+  end
+
+  describe 'ip validation' do
+    context 'when given valid ip addresses' do
+      subject do
+        [
+          '10.0.0.0',
+          '10.0.0.1',
+          '10.0.0.1:8080',
+          'FEDC:BA98:7654:3210:FEDC:BA98:7654:3210',
+          '1080:0000:0000:0000:0008:0800:200C:417A',
+          '1080:0:0:0:8:800:200C:417A',
+          '1080:0::8:800:200C:417A',
+          '1080::8:800:200C:417A',
+          'FF01:0:0:0:0:0:0:43',
+          'FF01:0:0::0:0:43',
+          'FF01::43',
+          '0:0:0:0:0:0:0:1',
+          '0:0:0::0:0:1',
+          '::1',
+          '0:0:0:0:0:0:0:0',
+          '0:0:0::0:0:0',
+          '::',
+          'fe80::208:74ff:feda:625c',
+          'fe80::208:74ff:feda:625c%eth0',
+          'ff80:03:02:01::',
+          '[fe80::208:74ff:feda:625c]',
+          '[fe80::208:74ff:feda:625c]:8080',
+          '[fe80::208:74ff:feda:625c%eth0]:8080'
+        ]
+      end
+
+      it { is_expected.to all(be_valid_ip) }
+    end
+
+    context 'when given invalid ip addresses' do
+      subject do
+        [
+          '',
+          '10.0.0.256',
+          '10.0.0.0.0',
+          '10.0.0',
+          '10.0',
+          '0.0.0.0/0',
+          '10.0.0.1/24',
+          '10.0.0.1/255.255.255.0',
+          ':1:2:3:4:5:6:7',
+          ':1:2:3:4:5:6:7',
+          '2002:516:2:200',
+          'dd',
+          '2001:db8::8:800:200c:417a/64',
+          '02001:0000:1234:0000:0000:C1C0:ABCD:0876',
+          '2001:0000:1234:0000:00001:C1C0:ABCD:0876'
+        ]
+      end
+
+      it { is_expected.to_not include(be_valid_ip) }
+    end
+  end
+
+  describe '#set_client_ip_tag' do
+    let(:span) do
+      instance_double('Span')
+    end
+
+    context 'when disabled' do
+      let(:tagging_enabled) { false }
+
+      it 'does nothing' do
+        expect(span).to_not receive(:set_tag).with(any_args)
```

Comment:
`with(any_args)` is redundant and assertion `expect(span).to_not receive(:set_tag)` should be strong enough.

Reply:
Awesome, fixed